src/bpf/bpfctl.c

```diff
@@ -0,0 +1,238 @@
+        if (!map_table)
+                return log_oom();
+
+        prog_table = table_new("program_name", "program_type", "memlock", "program_tag", "program_map_names", "run_time_ns", "run_cnt");
+        if (!prog_table)
+                return log_oom();
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_error_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
+
+        r = varlink_collect_full(vl, "io.systemd.Bpf.Describe", i, &o, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to collect varlink method: %m");
+
+        JSON_VARIANT_ARRAY_FOREACH(i, o) {
+                m = json_variant_by_key(i, "type");
+                if (!m)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to find type field of json variant");
+                v = json_variant_by_key(i, "data");
+                if (!v)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to find data field of json variant");
+
+                if (streq(json_variant_string(m), "map")) {
+                        static const JsonDispatch dispatch_table[] = {
+                                { "name",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfMapData, name),  JSON_MANDATORY },
+                                { "type", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, type), JSON_MANDATORY},
+                                { "memlock", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, memlock), JSON_MANDATORY},
+                                {}
+                        };
+                        _cleanup_(bpf_map_data_done) BpfMapData p = {};
+
+                        r = json_dispatch(v, dispatch_table, 0, &p);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to dispatch map table: %m");
+
+                        r = table_add_many(map_table,
+                                   TABLE_STRING, p.name, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.type, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.memlock, TABLE_SET_ALIGN_PERCENT, 100);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to add to map table: %m");
+                } else {
+                        static const JsonDispatch dispatch_table[] = {
+                                { "name",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, name),  JSON_MANDATORY },
+                                { "type", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, type), JSON_MANDATORY},
+                                { "memlock", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, memlock), JSON_MANDATORY},
```

Comment:
here and everywhere: when you use json_dispatch_uint64() then please set the 2nd column here to: _JSON_VARIANT_TYPE_INVALID. That has the effect that the field type check is dropped. This is a good idea, because JSON famously is not 64bit clean, because of its JS heritage. i.e. integers are only safe in the 53bit range. This means, when inserting integers in the 53…64 bit range various implementations format them as string field. 

json_dispatch_uint64() will handle this transparently for you, if you let it. But that only works, if you allow both integers and strings through. But via fixating the type to JSON_VARIANT_UNSIGNED here you do not allow that.

hence, whenever you use json_dispatch_uint64() (and similar number parsers) make sure to set the 2nd column to _JSON_VARIANT_TYPE_INVALID. json_dispatch_uint64() internally validates the types anyway, hence you lose nothing